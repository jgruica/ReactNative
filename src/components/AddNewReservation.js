import React, { Component } from 'react';
import {
  View,
  Text,
  TextInput,
  TouchableOpacity,
  Picker
} from 'react-native';
import moment from 'moment';
import DateTimePicker from 'react-native-modal-datetime-picker';
import Loader from './Loader';
import { graphql, compose, withApollo } from 'react-apollo';
import { createReservation } from '../../mutations/AllMutations';
import { reservations } from '../../queries/AllQueries';

import Icon from 'react-native-vector-icons/Entypo';

const styles = {
  wrapper: {
    flex: 1,
    paddingTop: 23,
    alignItems: 'stretch'
  },
  text: {
    fontSize: 15,
    textAlign: 'left',
    marginLeft: 10,
    color: '#2e3131'
  },
  input: {
    margin: 10,
    height: 40,
    borderColor: '#8c14fc',
    borderWidth: 1
  },
  submitButton: {
    backgroundColor: '#8c14fc',
    padding: 10,
    margin: 15,
    height: 40,
  },
  submitButtonText: {
    color: 'white'
  }
}

class AddNewReservation extends Component {
  state = {
    name: '',
    hotelName: '',
    arrivalDate: '',
    departureDate: '',
    creatingReservation: false,
    creatingReservationErr: "",
    arrivalDatePickerVisible: false,
    departureDatePickerVisible: false,
  }

  createNewReservation = () => {
    const { name, hotelName, arrivalDate, departureDate } = this.state;
    if (!name.length || !hotelName.length || !arrivalDate || !departureDate) {
      this.setState({ creatingReservationErr: "Fields can't be empty", creatingReservation: false });
      return;
    }
    this.setState({ creatingReservation: true });
    this.props.createReservation({ name, hotelName, arrivalDate, departureDate })
      .then(() => {
        this.setState({ creatingReservation: false });
        this.props.history.push('/');
      })
      .catch(() => {
        this.setState({ creatingReservation: false });
      });
  }

  showDateTimePicker = (field) => {
    this.setState({ [field]: true });
  }

  hideDateTimePicker = (field) => {
    this.setState({ [field]: false });
  }

  handleDatePicked = (date, dateField, hideField) => {
    this.setState({
      [dateField]: date
    })
    this.hideDateTimePicker(hideField);
  };


  render() {
    const { creatingReservationErr, creatingReservation } = this.state;
    const { data } = this.props;
    
    const arrivalDateValue = this.state.arrivalDate !== '' ?
      moment(this.state.arrivalDate).format('l').toString() : '';
    const departureDateValue = this.state.departureDate !== '' ?
      moment(this.state.departureDate).format('l').toString() : '';


    return (
      <View style={styles.wrapper}>
        {creatingReservationErr ? <Text>{creatingReservationErr}</Text> : null}
        <Text style={styles.text}>Name:</Text>
        <TextInput
          style={styles.input}
          placeholder="Name"
          onChangeText={(name) => this.setState({ name })} />
        <Text style={styles.text}>Pick hotel: </Text>
        {data.loading ? 
         <Loader loading={data.loading} />
        :
        <Picker
        selectedValue={this.state.hotelName}
        onValueChange={(itemValue, itemIndex) =>
          this.setState({ hotelName: itemValue })
        }>
        {data.reservations.map(reservation => {
          return (
            <Picker.Item label={reservation.hotelName} value={reservation.hotelName} />
          )
        })}
      </Picker>
        }
        <TouchableOpacity onPress={() => this.showDateTimePicker('arrivalDatePickerVisible')}>
          <TextInput
            style={styles.input}
            placeholder="Choose arrival time"
            value={arrivalDateValue}
            editable={false}
            pointerEvents="none"
          />
        </TouchableOpacity>
        <DateTimePicker
          isVisible={this.state.arrivalDatePickerVisible}
          mode='date'
          onConfirm={(date) => this.handleDatePicked(date, 'arrivalDate', 'arrivalDatePickerVisible')}
          onCancel={() => this.hideDateTimePicker('arrivalDatePickerVisible')}
        />
        <TouchableOpacity onPress={() => this.showDateTimePicker('departureDatePickerVisible')}>
          <TextInput
            style={styles.input}
            placeholder="Choose departure time"
            value={departureDateValue}
            editable={false}
            pointerEvents="none"
          />
        </TouchableOpacity>
        <DateTimePicker
          isVisible={this.state.departureDatePickerVisible}
          mode='date'
          onConfirm={(date) => this.handleDatePicked(date, 'departureDate', 'departureDatePickerVisible')}
          onCancel={() => this.hideDateTimePicker('departureDatePickerVisible')}
        />

        <TouchableOpacity
          style={styles.submitButton}
          onPress={this.createNewReservation}
        >
          <Text style={styles.submitButtonText}>{!creatingReservation ? 'Add reservation' : 'Adding...'} </Text>
        </TouchableOpacity>
      </View>
    );
  }
}

export default compose(
  withApollo,
  graphql(createReservation, {
    props: ({ mutate }) => ({
      createReservation: (data) => mutate({ variables: { data } })
    })
  }),
  graphql(reservations, {
    options: {
      fetchPolicy: "network-only"
    }
  }),
)(AddNewReservation);