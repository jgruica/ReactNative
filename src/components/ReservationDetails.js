import React, { Component } from 'react';
import {
  Text,
  TextInput,
  View,
  TouchableOpacity,
  Modal,
  TouchableHighlight
} from 'react-native';
import { graphql, compose, withApollo } from 'react-apollo';

import moment from 'moment';
import DateTimePicker from 'react-native-modal-datetime-picker';

import Loader from "./Loader";
import { reservationQuery } from '../../queries/AllQueries';
import { updateReservation, deleteReservation } from '../../mutations/AllMutations';

import styles from "../../styles/ReservationDetailsStyle";

export class ReservationDetails extends Component {
  state = {
    name: '',
    hotelName: '',
    arrivalDate: '',
    departureDate: '',
    displayDeleteModal: false,
    updatingReservation: false,
    updatingReservationErr: '',
    arrivalDatePickerVisible: false,
    departureDatePickerVisible: false,
  }

  static navigationOptions = {
    title: "Details",
  }

  toggleDeleteModal = () => {
    this.setState({ displayDeleteModal: !this.state.displayDeleteModal })
  }

  onPressDelete = () => {
    this.toggleDeleteModal();
  }

  componentDidMount() {
    if (this.props.data.reservation) {
      this.setState({
        name: this.props.data.reservation.name,
        hotelName: this.props.data.reservation.hotelName,
        arrivalDate: this.props.data.reservation.arrivalDate,
        departureDate: this.props.data.reservation.departureDate
      })
    }
  }

  componentDidUpdate(lastProps) {
    if (this.props.data.reservation !== lastProps.data.reservation && this.props.data.reservation) {
      this.setState({
        name: this.props.data.reservation.name,
        hotelName: this.props.data.reservation.hotelName,
        arrivalDate: this.props.data.reservation.arrivalDate,
        departureDate: this.props.data.reservation.departureDate
      })
    }
  }

  updateReservation = () => {
    const { name, hotelName, arrivalDate, departureDate } = this.state;
    if (!name.length || !hotelName.length || !arrivalDate.length || !departureDate.length) {
      this.setState({ updatingReservationErr: "Fields can't be empty", updatingReservation: false });
      return;
    }
    this.setState({ updatingReservation: true });
    this.props.updateReservation({ name, hotelName, arrivalDate, departureDate }, { id: this.props.navigation.getParam('id') })
      .then(() => {
        this.setState({ updatingReservation: false });
        this.props.navigation.navigate('Home');
      })
      .catch(() => {
        this.setState({ updatingReservation: false });
      });
  }

  deleteReservation = () => {
    this.props.deleteReservation({ id: this.props.navigation.getParam('id') })
      .then(() => {
        this.toggleDeleteModal();
        this.props.navigation.navigate('Home');
      })
      .catch();
  }

  showDateTimePicker = (field) => {
    this.setState({ [field]: true });
  }

  hideDateTimePicker = (field) => {
    this.setState({ [field]: false });
  }

  handleDatePicked = (date, dateField, hideField) => {
    this.setState({
      [dateField]: date
    })
    this.hideDateTimePicker(hideField);
  };

  render() {
    const { updatingReservationErr } = this.state;
    const { data } = this.props;

    const arrivalDateValue = this.state.arrivalDate !== '' ?
      moment(this.state.arrivalDate).format('l').toString() : '';
    const departureDateValue = this.state.departureDate !== '' ?
      moment(this.state.departureDate).format('l').toString() : '';

    return (
      <View style={styles.wrapper}>
        {data.loading ?
          <Loader loading={data.loading} />
          :
          <View style={{ padding: 5 }}>
            {updatingReservationErr ? <Text>{updatingReservationErr}</Text> : null}
            <Text style={styles.text}>Name:</Text>
            <TextInput
              style={styles.input}
              defaultValue={data.reservation.name}
              onChangeText={(name) => this.setState({ name })} />
            <Text style={styles.text}>Hotel name:</Text>
            <TextInput
              style={styles.input}
              defaultValue={data.reservation.hotelName}
              onChangeText={(hotelName) => this.setState({ hotelName })} />
            <Text style={styles.text}>Arrival time:</Text>
            <TouchableOpacity onPress={() => this.showDateTimePicker('arrivalDatePickerVisible')}>
              <TextInput
                style={styles.input}
                defaultValue={arrivalDateValue}
                editable={false}
                pointerEvents="none"
              />
            </TouchableOpacity>
            <DateTimePicker
              isVisible={this.state.arrivalDatePickerVisible}
              mode='date'
              onConfirm={(date) => this.handleDatePicked(date, 'arrivalDate', 'arrivalDatePickerVisible')}
              onCancel={() => this.hideDateTimePicker('arrivalDatePickerVisible')}
            />
              <Text style={styles.text}>Departure time:</Text>
            <TouchableOpacity onPress={() => this.showDateTimePicker('departureDatePickerVisible')}>
              <TextInput
                style={styles.input}
                defaultValue={departureDateValue}
                editable={false}
                pointerEvents="none"
              />
            </TouchableOpacity>
            <DateTimePicker
              isVisible={this.state.departureDatePickerVisible}
              mode='date'
              onConfirm={(date) => this.handleDatePicked(date, 'departureDate', 'departureDatePickerVisible')}
              onCancel={() => this.hideDateTimePicker('departureDatePickerVisible')}
            />
          </View>
        }
        <View style={styles.buttonsContainer}>
        <TouchableOpacity
            style={styles.deleteButton}
            onPress={this.onPressDelete}
          >
            <Text style={styles.buttonText}>Delete</Text>
          </TouchableOpacity>
          <TouchableOpacity
            style={styles.updateButton}
            onPress={this.updateReservation}
          >
            <Text style={styles.updateButtonText}>Update</Text>
          </TouchableOpacity>
        </View>
        {this.state.displayDeleteModal &&
          <Modal
            animationType="fade"
            transparent={true}
            visible={this.state.displayDeleteModal}
            onRequestClose={() => {
              Alert.alert('Modal has been closed.');
            }}>
            <View style={styles.modalContent}>
              <View>
                <Text style={styles.text}>Are you sure you want to delete this reservation? </Text>
                <View style={styles.buttonsContainer}>
                  <TouchableHighlight
                    style={styles.cancelButon}
                    onPress={() => {
                      this.toggleDeleteModal();
                    }}>
                    <Text style={styles.cancelText}>Cancel</Text>
                  </TouchableHighlight>
                  <TouchableHighlight
                    style={styles.deleteButton}
                    onPress={this.deleteReservation}>
                    <Text style={styles.buttonText}>Delete</Text>
                  </TouchableHighlight>
                </View>
              </View>
            </View>
          </Modal>
        }
      </View>

    )
  }
}


export default compose(
  withApollo,
  graphql(reservationQuery, {
    options: ({ navigation }) => ({
      variables: { where: { id: navigation.getParam('id')} }
    })
  }),
  graphql(updateReservation, {
    props: ({ mutate }) => ({
      updateReservation: (data, where) => mutate({ variables: { data, where } })
    })
  }),
  graphql(deleteReservation, {
    props: ({ mutate }) => ({
      deleteReservation: (where) => mutate({ variables: { where } })
    })
  }),
)(ReservationDetails);