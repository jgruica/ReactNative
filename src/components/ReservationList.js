import React, { Component } from 'react';
import {
  Text,
  View,
  ScrollView,
  TouchableOpacity
} from 'react-native';
import { graphql, compose, withApollo } from 'react-apollo';
import Icon from 'react-native-vector-icons/Entypo';

import moment from 'moment';
import sortBy from 'lodash/sortBy';

import Loader from './Loader';

import { reservations } from '../../queries/AllQueries';

import styles from "../../styles/ReservationListStyle";

const colors = ['#f1a9a0', '#d2527f', '#e74c3c', '#9f5afd', 
'#913d88', '#81cfe0', '#1e8bc3', '#2ecc71', '#1e824c', '#fef160', '#e67e22', '#f5e51b', '#f27935']
let nextColorIndex = 0
const colorMap = {}
const getHotelColor = (hotelName) => {
  if (!colorMap[hotelName]) {
    colorMap[hotelName] = colors[nextColorIndex]
    nextColorIndex = (nextColorIndex + 1) % colors.length
  }
  return {
    flex: 1,
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
    borderBottomColor: '#A9A9A9',
    borderBottomWidth: 0.2,
    borderLeftColor: colorMap[hotelName],
    borderLeftWidth: 10,
  }
}

class ReservationList extends Component {
  static navigationOptions = {
    title: "Reservations",
  }
  render() {
    const { data } = this.props;
    const sortedReservations = data.reservations && data.reservations.length && sortBy(data.reservations, "hotelName");
    const detailsIcon = <Icon name="chevron-small-right" size={30} color="#A9A9A9" />;
    if (data.loading) {
      return <Loader loading={data.loading} />;
    } else {
      return (
        <View>
          <ScrollView>
            {sortedReservations.map(reservation =>
              <View key={reservation.id} style={getHotelColor(reservation.hotelName)} >
                <Text style={styles.header}>{reservation.name}</Text>
                <Text style={styles.text}>{moment(reservation.arrivalDate).format("ll")} - {moment(reservation.departureDate).format("ll")}</Text>
                <Text style={styles.text}>{reservation.hotelName}</Text>
                <TouchableOpacity onPress={() =>
                  this.props.navigation.navigate('Details', { id: reservation.id })}
                  style={{ position: 'absolute', top: 10, right: 20 }}>
                  <Text>{detailsIcon}</Text>
                </TouchableOpacity>
              </View>
            )}
          </ScrollView>
          <TouchableOpacity onPress={() =>
            this.props.navigation.navigate('AddNew')
          } style={{
            width: 60, height: 60, borderRadius: 60 / 2, backgroundColor: '#8c14fc',
            position: 'absolute', bottom: 30, right: 20, alignSelf: 'flex-end'
          }} underlayColor="#f0f4f7">
            <Text style={{ fontSize: 25, color: '#FFFFFF', left: 22, top: 12 }}>+</Text>
          </TouchableOpacity>
        </View>
      );
    }
  }
}

export default compose(
  withApollo,
  graphql(reservations, {
    options: {
      fetchPolicy: "network-only"
    }
  }),
)(ReservationList);