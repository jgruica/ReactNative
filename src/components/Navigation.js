import { createStackNavigator, createAppContainer } from 'react-navigation';
import ReservationList from './components/ReservationList';
import ReservationDetails from './components/ReservationDetails';
import AddNewReservation from './components/AddNewReservation';

const Root = createStackNavigator({
    Home: { 
      screen: ReservationList,
    },
    Details: { 
      screen: ReservationDetails,
    },
    AddNew: {
      screen: AddNewReservation
    }
});

const Navigation = createAppContainer(Root);
export default Navigation; 