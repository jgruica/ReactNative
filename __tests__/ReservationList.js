import 'jsdom-global/register';
import React from 'react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import  { ReservationList } from '../src/components/ReservationList';

configure({ adapter: new Adapter() });
import renderer from 'react-test-renderer';

describe("ReservationList", () => {
  const data = {
    loading: false,
    error: null,
    reservations: [{
      id: '123',
      name: 'Test Name',
      hotelName: 'Test Hotel',
      arrivalDate: '01/01/2020',
      departureDate: '01/05/2020'
    }]
  }

  it('renders Reservation List', () => {
    const wrapper = renderer.create(
      <ReservationList 
        data={data}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
