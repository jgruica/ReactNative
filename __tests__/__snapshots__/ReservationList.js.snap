// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReservationList renders Reservation List 1`] = `
<View>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "borderBottomColor": "#A9A9A9",
            "borderBottomWidth": 0.2,
            "borderLeftColor": "#f1a9a0",
            "borderLeftWidth": 10,
            "flex": 1,
            "justifyContent": "flex-start",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2e3131",
              "fontSize": 15,
              "marginLeft": 10,
              "textAlign": "left",
            }
          }
        >
          Test Name
        </Text>
        <Text
          style={
            Object {
              "color": "#A9A9A9",
              "fontSize": 15,
              "marginLeft": 10,
              "textAlign": "left",
            }
          }
        >
          Jan 1, 2020
           - 
          Jan 5, 2020
        </Text>
        <Text
          style={
            Object {
              "color": "#A9A9A9",
              "fontSize": 15,
              "marginLeft": 10,
              "textAlign": "left",
            }
          }
        >
          Test Hotel
        </Text>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "position": "absolute",
              "right": 20,
              "top": 10,
            }
          }
        >
          <Text>
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#A9A9A9",
                    "fontSize": 30,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Entypo",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              ï…†
            </Text>
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignSelf": "flex-end",
        "backgroundColor": "#8c14fc",
        "borderRadius": 30,
        "bottom": 30,
        "height": 60,
        "opacity": 1,
        "position": "absolute",
        "right": 20,
        "width": 60,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 25,
          "left": 22,
          "top": 12,
        }
      }
    >
      +
    </Text>
  </View>
</View>
`;
